### PR TITLE
428: Add language header to po exports

### LINKED ISSUE #428: Add header "Language" to metadata  .po
Exported po-files incompatible with some libraries.
meta Language needed.
You can easy fix it.

file:
gp-includes/formats/format_pomo.php

add at line: 20
`$po->set_header( 'Language', $locale->lang_code_iso_639_1 );
` 

result:

```
        $po->set_header( 'Content-Transfer-Encoding', '8bit' );
        $po->set_header( 'Plural-Forms', "nplurals=$locale->nplurals; plural=$locale->plural_expression;" );
        $po->set_header( 'X-Generator', 'GlotPress/' . GP_VERSION );
        $po->set_header( 'Language', $locale->lang_code_iso_639_1 );  // new line !!!

        // force export only current translations
        $filters = array();

```

### PR DESCRIPTION
**Pull request by toolstack**
Fixes #428.

### CODE REVIEW

File: gp-includes/formats/format_pomo.php [hunk 1]
```diff
@@ -18,7 +19,12 @@ public function print_exported_file( $project, $locale, $translation_set, $entri
 		$po->set_header( 'Plural-Forms', "nplurals=$locale->nplurals; plural=$locale->plural_expression;" );
 		$po->set_header( 'X-Generator', 'GlotPress/' . GP_VERSION );
 
-		// force export only current translations
+		$language_string = $this->generate_language_string( $locale );
+		if ( false !== $language_string ) {
+			$po->set_header( 'Langauge', $language_string );
```
Comment (ocean90) [inline gp-includes/formats/format_pomo.php]: Typo, Langauge => Language

File: gp-includes/formats/format_pomo.php [hunk 2]
```diff
@@ -45,6 +51,36 @@ public function read_originals_from_file( $file_name ) {
...
+	 * @param GP_Locale $locale The locale object.
+	 *
+	 * @return string|false
+	 */
+	private function generate_language_string( $locale ) {
+		$ret = '';
+
+		if ( $locale->lang_code_iso_639_1 ) {
+			$ret = $locale->lang_code_iso_639_1;
+		} else if ( $locale->lang_code_iso_639_2 ) {
```
Comment (ocean90) [inline gp-includes/formats/format_pomo.php]: Should be `elseif`, see https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#use-elseif-not-else-if.

File: gp-includes/formats/format_pomo.php [hunk 3]
```diff
@@ -45,6 +51,36 @@ public function read_originals_from_file( $file_name ) {
...
+		} else if ( $locale->lang_code_iso_639_3 ) {
+			$ret = $locale->lang_code_iso_639_3;
+		}
+
+		if ( '' === $ret ) {
+			return false;
+		}
+
+		if ( null !== $locale->country_code && strcasecmp( $ret, $locale->country_code ) === 0 ) {
+			$ret .= '_' . $locale->country_code;
```
Comment (ocean90) [inline gp-includes/formats/format_pomo.php]: In `GP_Locales` the country code is lowercase, but the specification says "‘ll_CC’, where ‘ll’ is an ISO 639 two-letter language code (lowercase) and **‘CC’ is an ISO 3166 two-letter country code (uppercase).**":
`$ret .= '_' . strtoupper( $locale->country_code )`

File: gp-includes/formats/format_pomo.php [hunk 4]
```diff
@@ -45,6 +51,36 @@ public function read_originals_from_file( $file_name ) {
...
+	/**
+	 * Create a string that represents the value for the "Language:" header for a po file.
+	 *
+	 * @since 2.1.0
+	 *
+	 * @param GP_Locale $locale The locale object.
+	 *
+	 * @return string|false
+	 */
+	private function generate_language_string( $locale ) {
```
Comment (ocean90) [inline gp-includes/formats/format_pomo.php]: Can we name this `get_language_code()` instead?

File: gp-includes/formats/format_pomo.php [hunk 5]
```diff
@@ -45,6 +51,36 @@ public function read_originals_from_file( $file_name ) {
...
+			$ret = $locale->lang_code_iso_639_2;
+		} elseif ( $locale->lang_code_iso_639_3 ) {
+			$ret = $locale->lang_code_iso_639_3;
+		}
+
+		if ( '' === $ret ) {
+			return false;
+		}
+
+		if ( null !== $locale->country_code && strcasecmp( $ret, $locale->country_code ) === 0 ) {
```
Comment (akirk) [inline gp-includes/formats/format_pomo.php]: Maybe it's just me but I don't get this (especially the second part). Is this for something like de_DE? what about de_CH? Worth a comment and/or unit test?

Comment (toolstack) [inline gp-includes/formats/format_pomo.php]: Yes, 'de_DE' == 'de' so the code skips adding it if they're the same.  When they're not the same (like de_CH) it gets added.

Comment (akirk) [inline gp-includes/formats/format_pomo.php]: Shouldn't it then be `!== 0`?

Comment (toolstack) [inline gp-includes/formats/format_pomo.php]: Yes... yes it should :wink:

File: gp-includes/formats/format_pomo.php [hunk 6]
```diff
@@ -18,7 +19,12 @@ public function print_exported_file( $project, $locale, $translation_set, $entri
 		$po->set_header( 'Plural-Forms', "nplurals=$locale->nplurals; plural=$locale->plural_expression;" );
 		$po->set_header( 'X-Generator', 'GlotPress/' . GP_VERSION );
 
-		// force export only current translations
+		$language_string = $this->get_language_code( $locale );
```
Comment (akirk) [inline gp-includes/formats/format_pomo.php]: Nitpick but also change the variable name along with the function name change?

File: gp-includes/formats/format_pomo.php [hunk 7]
```diff
@@ -45,6 +51,36 @@ public function read_originals_from_file( $file_name ) {
...
 	}
 
+	/**
+	 * Create a string that represents the value for the "Language:" header for a po file.
+	 *
+	 * @since 2.1.0
+	 *
+	 * @param GP_Locale $locale The locale object.
+	 *
+	 * @return string|false
```
Comment (ocean90) [inline gp-includes/formats/format_pomo.php]: `@return` should have a description to explain when it returns `false`.

File: gp-includes/formats/format_pomo.php [hunk 8]
```diff
@@ -45,6 +51,36 @@ public function read_originals_from_file( $file_name ) {
...
+			$ret = $locale->lang_code_iso_639_2;
+		} elseif ( $locale->lang_code_iso_639_3 ) {
+			$ret = $locale->lang_code_iso_639_3;
+		}
+
+		if ( '' === $ret ) {
+			return false;
+		}
+
+		if ( null !== $locale->country_code && strcasecmp( $ret, $locale->country_code ) !== 0 ) {
```
Comment (ocean90) [inline gp-includes/formats/format_pomo.php]: The `strcasecmp()` check should be a yoda condition too. (https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#yoda-conditions)

**Comment by ocean90** [comment]
A unit test to make codecov happy would be great.

**Comment by toolstack** [comment]
@ocean90 I didn't add a unit test as there are no unit tests for export at the moment.

**Comment by toolstack** [comment]
Issue #437 has been opened to handle the test cases.

**Pull request merged**